File: app/src/main/java/se/gustavkarlsson/skylight/android/dagger/modules/EvaluationModule.java
package se.gustavkarlsson.skylight.android.dagger.modules;

import dagger.Binds;
import dagger.Module;
import dagger.Reusable;
import se.gustavkarlsson.skylight.android.evaluation.AuroraReportEvaluator;
import se.gustavkarlsson.skylight.android.evaluation.ChanceEvaluator;
import se.gustavkarlsson.skylight.android.evaluation.DarknessEvaluator;
import se.gustavkarlsson.skylight.android.evaluation.GeomagActivityEvaluator;
import se.gustavkarlsson.skylight.android.evaluation.GeomagLocationEvaluator;
import se.gustavkarlsson.skylight.android.evaluation.VisibilityEvaluator;
import se.gustavkarlsson.skylight.android.models.AuroraReport;
import se.gustavkarlsson.skylight.android.models.factors.Darkness;
import se.gustavkarlsson.skylight.android.models.factors.GeomagActivity;
import se.gustavkarlsson.skylight.android.models.factors.GeomagLocation;
import se.gustavkarlsson.skylight.android.models.factors.Visibility;

@Module
public abstract class EvaluationModule {

	@Binds
	@Reusable
	abstract ChanceEvaluator<GeomagActivity> bindGeomagActivityEvaluator(GeomagActivityEvaluator evaluator);

	@Binds
	@Reusable
	abstract ChanceEvaluator<GeomagLocation> bindGeomagLocationEvaluator(GeomagLocationEvaluator evaluator);

	@Binds
	@Reusable
	abstract ChanceEvaluator<Visibility> bindVisibilityEvaluator(VisibilityEvaluator evaluator);

	@Binds
	@Reusable
	abstract ChanceEvaluator<Darkness> bindDarknessEvaluator(DarknessEvaluator evaluator);

	@Binds
	@Reusable
	abstract ChanceEvaluator<AuroraReport> bindAuroraReportEvaluator(AuroraReportEvaluator evaluator);

}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/dagger/modules/definitive/EvaluationModule.kt
package se.gustavkarlsson.skylight.android.dagger.modules.definitive

import dagger.Binds
import dagger.Module
import dagger.Reusable
import se.gustavkarlsson.skylight.android.entities.*
import se.gustavkarlsson.skylight.android.services.evaluation.ChanceEvaluator
import se.gustavkarlsson.skylight.android.services_impl.evaluation.*

@Module
abstract class EvaluationModule {

    @Binds
    @Reusable
    abstract fun bindGeomagActivityEvaluator(impl: GeomagActivityEvaluator): ChanceEvaluator<GeomagActivity>

    @Binds
    @Reusable
    abstract fun bindGeomagLocationEvaluator(impl: GeomagLocationEvaluator): ChanceEvaluator<GeomagLocation>

    @Binds
    @Reusable
    abstract fun bindVisibilityEvaluator(impl: VisibilityEvaluator): ChanceEvaluator<Visibility>

    @Binds
    @Reusable
    abstract fun bindDarknessEvaluator(impl: DarknessEvaluator): ChanceEvaluator<Darkness>

    @Binds
    @Reusable
    abstract fun bindAuroraReportEvaluator(impl: AuroraReportEvaluator): ChanceEvaluator<AuroraReport>

}
